src/java.base/unix/native/libjava/TimeZone_md.c

```diff
@@ -555,58 +555,40 @@ findJavaTZ_md(const char *java_home_dir)
 /**
  * Returns a GMT-offset-based zone ID. (e.g., "GMT-08:00")
  */
-
-#if defined(MACOSX)
-
 char *
 getGMTOffsetID()
 {
-    time_t offset;
-    char sign, buf[32];
-    struct tm local_tm;
-    time_t clock;
-
-    clock = time(NULL);
-    if (localtime_r(&clock, &local_tm) == NULL) {
+    char buf[32];
+    char offset[6];
+    struct tm localtm;
+    time_t clock = time(NULL);
+    if (localtime_r(&clock, &localtm) == NULL) {
         return strdup("GMT");
     }
-    offset = (time_t)local_tm.tm_gmtoff;
-    if (offset == 0) {
+
+#if defined(MACOSX)
+    time_t gmt_offset;
+    gmt_offset = (time_t)localtm.tm_gmtoff;
+    if (gmt_offset == 0) {
         return strdup("GMT");
     }
-    if (offset > 0) {
-        sign = '+';
-    } else {
-        offset = -offset;
-        sign = '-';
-    }
-    sprintf(buf, (const char *)"GMT%c%02d:%02d",
-            sign, (int)(offset/3600), (int)((offset%3600)/60));
-    return strdup(buf);
-}
-
 #else
+    struct tm gmt;
+    if (gmtime_r(&clock, &gmt) == NULL) {
+        return strdup("GMT");
+    }
 
-char *
-getGMTOffsetID()
-{
-    time_t offset;
-    char sign, buf[32];
-    offset = timezone;
-
-    if (offset == 0) {
+    if(localtm.tm_hour == gmt.tm_hour && localtm.tm_min == gmt.tm_min) {
         return strdup("GMT");
     }
+#endif
 
-    /* Note that the time offset direction is opposite. */
-    if (offset > 0) {
-        sign = '-';
-    } else {
-        offset = -offset;
-        sign = '+';
+    strftime(offset, 6, "%z", &localtm);
+    if (strlen(offset) != 5) {
```

Comment:
What I meant was the return value from `strftime`, not the length from `strlen`, ie, `if (strftime(...) != 5)`. It should be more robust.